src/cart/src/include/gurt/debug_setup.h

```diff
@@ -0,0 +1,289 @@
+		int __rc = 0;						\
+		do {							\
+			db_foreach(_D_LOG_DEALLOCATE_DBG_BIT, __rc)	\
+		} while (0);						\
+		__rc;							\
+	})
+
+/** Declare an enumeration to identify a debug flag, facility combination
+ *
+ * \param flag		The name of the variable used to store the value of the
+ *			debug bit.
+ * \param s_name	Unused
+ * \param l_name	Unused
+ * \param mask		Unused
+ * \param fac		The log facility
+ */
+#define _D_LOG_DECLARE_ENUM(flag, s_name, l_name, mask, fac)	\
+	DD_FLAG_NAME(flag, fac),
+
+/** Internal macro for initializing facility cache defined by DD_*_CACHE */
+#define _D_LOG_INITIALIZE_FIELD(flag, s_name, l_name, mask, fac) DLOG_UNINIT,
+
+/**
+ * These macros are intended to be used with FOREACH macros that define log
+ * facilities in your library.   It will utilize internal FOREACH macros to
+ * define debug bits as well as user defined macros.  See D_FOREACH_GURT_FAC for
+ * an example.
+ *
+ * These macros can also be used standalone but are not quite as convenient due
+ * to required but unused arguments
+ */
+
+/** Internal macro to declare a facility cache.  Used by DD_DECLARE_FAC */
+#define _D_LOG_DECLARE_CACHE(s_name, user_dbg_bits)			\
+	enum {								\
+		DD_FLAG_NAME(s_name, START),				\
+		D_FOREACH_PRIO_MASK(_D_LOG_DECLARE_ENUM, DD_FAC(s_name))\
+		D_FOREACH_GURT_DB(_D_LOG_DECLARE_ENUM, DD_FAC(s_name))	\
+		user_dbg_bits(_D_LOG_DECLARE_ENUM, DD_FAC(s_name))	\
+		DD_FLAG_NAME(s_name, END),				\
+	};								\
+	extern int DD_CACHE(DD_FAC(s_name))[DD_FLAG_NAME(s_name, END) - 1];
+
+/** Internal macro to create a facility cache.  Used by DD_INSTANTIATE_FAC */
+#define _D_LOG_INSTANTIATE_CACHE(s_name, user_dbg_bits)			\
+	int DD_CACHE(DD_FAC(s_name))[DD_FLAG_NAME(s_name, END) - 1] = {	\
+		D_FOREACH_PRIO_MASK(_D_LOG_INITIALIZE_FIELD, D_NOOP)	\
+		D_FOREACH_GURT_DB(_D_LOG_INITIALIZE_FIELD, D_NOOP)	\
+		user_dbg_bits(_D_LOG_INITIALIZE_FIELD, D_NOOP)		\
+	};
+
+/** Internal macro to register a facility cache.  Used by
+ *  ALLOCATE_LOG_FACILITY
+ */
+#define _D_ADD_CACHE(s_name)					\
+	d_log_add_cache(DD_CACHE(DD_FAC(s_name)),		\
+			ARRAY_SIZE(DD_CACHE(DD_FAC(s_name))));
+
+/** Declare an extern to global facility variable
+ *
+ * \param s_name	Short name for the facility
+ * \param l_name	Unused
+ * \param user_dbg_bits	Unused
+ */
+#define D_LOG_DECLARE_FAC(s_name, l_name, user_dbg_bits)	\
+	extern int DD_FAC_DECL(s_name);				\
+	_D_LOG_DECLARE_CACHE(s_name, user_dbg_bits)
+
+/** Declare a global facility variable
+ *
+ * \param s_name	Short name for the facility
+ * \param l_name	Unused
+ * \param user_dbg_bits	Unused
+ */
+#define D_LOG_INSTANTIATE_FAC(s_name, l_name, user_dbg_bits)	\
+	int DD_FAC_DECL(s_name);				\
+	_D_LOG_INSTANTIATE_CACHE(s_name, user_dbg_bits)
+
+/** Internal use macro as part of DD_REGISTER_FAC.  It's multiline on purpose
+ *  so we can break from the foreach loop.  Checkpatch will complain but don't
+ *  see a better way.
+ */
+#define _D_LOG_ALLOCATE_LOG_FACILITY(s_name, l_name, rc)		\
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon